src/extensions/builder.rs

```diff
@@ -0,0 +1,130 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! [`DftSessionStateBuilder`] for configuring DataFusion [`SessionState`]
+
+use datafusion::catalog::TableProviderFactory;
+use datafusion::execution::context::SessionState;
+use datafusion::execution::runtime_env::RuntimeEnv;
+use datafusion::execution::session_state::SessionStateBuilder;
+use datafusion::prelude::SessionConfig;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+/// Builds a DataFusion [`SessionState`] with any necessary configuration
+///
+/// Ideally we would use the DataFusion [`SessionStateBuilder`], but it doesn't
+/// currently have all the needed APIs. Once we have a good handle on the needed
+/// APIs we can upstream them to DataFusion.
+///
+/// List of things that would be nice to add upstream:
+/// TODO: Implement Debug for SessionStateBuilder upstream
+/// TODO: Implement with_table_factory to add a single TableProviderFactory to the list of factories
+/// TODO: Implement some way to get access to the current RuntimeEnv (to register object stores)
+/// TODO: Implement a way to add just a single TableProviderFactory
+/// TODO: Make TableFactoryProvider implement Debug
+/// TODO: rename RuntimeEnv::new() to RuntimeEnv::try_new() as it returns a Result
+//#[derive(Debug)]
+pub struct DftSessionStateBuilder {
+    session_config: SessionConfig,
+    table_factories: Option<HashMap<String, Arc<dyn TableProviderFactory>>>,
+    runtime_env: Option<Arc<RuntimeEnv>>,
+}
+
+impl Debug for DftSessionStateBuilder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DftSessionStateBuilder")
+            .field("session_config", &self.session_config)
+            //.field("table_factories", &self.table_factories)
+            .field(
+                "table_factories",
+                &"TODO TableFactoryDoes not implement Debug",
+            )
+            .field("runtime_env", &self.runtime_env)
+            .finish()
+    }
+}
+
+impl Default for DftSessionStateBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DftSessionStateBuilder {
+    /// Create a new builder
+    pub fn new() -> Self {
+        let session_config = SessionConfig::default()
+            // TODO why is batch size 1?
```

Comment:
I just copied this batch size of 1 here, but it seems quite inefficient

@matthewmturner  do you know why dft is using a batch size of 1?

Reply:
That was copied over from the Polygon terminal app i was working on that was querying web sockets where it was useful to test with batch size of 1.  We can definitely change that.